source/containerimage/ocilayout.go

```diff
@@ -0,0 +1,195 @@
+	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	sessionID := r.sessionID
+
+	caller, err := r.sm.Get(timeoutCtx, sessionID, false)
+	if err != nil {
+		return r.fetchWithAnySession(ctx, desc)
+	}
+
+	return r.fetchWithSession(ctx, desc, caller)
+}
+
+func (r *OCILayoutResolver) fetchWithAnySession(ctx context.Context, desc ocispecs.Descriptor) (io.ReadCloser, error) {
+	var rc io.ReadCloser
+	err := r.sm.Any(ctx, r.g, func(ctx context.Context, _ string, caller session.Caller) error {
+		readCloser, err := r.fetchWithSession(ctx, desc, caller)
+		if err != nil {
+			return err
+		}
+		rc = readCloser
+		return nil
+	})
+	return rc, err
+}
+
+func (r *OCILayoutResolver) fetchWithSession(ctx context.Context, desc ocispecs.Descriptor, caller session.Caller) (io.ReadCloser, error) {
+	store := sessioncontent.NewCallerStore(caller, fmt.Sprintf("oci-layout:%s", r.path))
+	readerAt, err := store.ReaderAt(ctx, desc)
+	if err != nil {
+		return nil, err
+	}
+	// just wrap the ReaderAt with a Reader
+	return io.NopCloser(&readerAtWrapper{readerAt: readerAt}), nil
+}
+
+// Resolve attempts to resolve the reference into a name and descriptor.
+// OCI Layout does not (yet) support tag name references, but does support hash references.
+func (r *OCILayoutResolver) Resolve(ctx context.Context, refString string) (string, ocispecs.Descriptor, error) {
+	ref, err := reference.Parse(refString)
+	if err != nil {
+		return "", ocispecs.Descriptor{}, fmt.Errorf("invalid reference '%s': %v", refString, err)
+	}
+	var (
+		dig  digest.Digest
+		info content.Info
+		rc   io.ReadCloser
+	)
+	if dig = ref.Digest(); dig == "" {
+		return "", ocispecs.Descriptor{}, fmt.Errorf("reference must have format @sha256:<hash>: %s", refString)
+	}
+	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	sessionID := r.sessionID
+
+	caller, err := r.sm.Get(timeoutCtx, sessionID, false)
+	if err != nil {
+		info, err = r.infoWithAnySession(ctx, dig)
+	} else {
+		info, err = r.infoWithSession(ctx, dig, caller)
+	}
+	if err != nil {
+		return "", ocispecs.Descriptor{}, fmt.Errorf("unable to get info about digest: %v", err)
+	}
+
+	// Create the descriptor, then use that to read the actual root manifest/
+	// This is necessary because we do not know the media-type of the descriptor,
+	// and there are descriptor processing elements that expect it.
+	desc := ocispecs.Descriptor{
+		Digest: dig,
+		Size:   info.Size,
+	}
+	caller, err = r.sm.Get(timeoutCtx, sessionID, false)
+	if err != nil {
+		rc, err = r.fetchWithAnySession(ctx, desc)
+	} else {
+		rc, err = r.fetchWithSession(ctx, desc, caller)
+	}
+	if err != nil {
+		return "", ocispecs.Descriptor{}, fmt.Errorf("unable to get root manifest: %v", err)
+	}
+	b, err := io.ReadAll(rc)
+	if err != nil {
+		return "", ocispecs.Descriptor{}, fmt.Errorf("unable to read root manifest: %v", err)
+	}
+	// try it first as an index, then as a manifest
```

Comment:
We have a `DetectManifestMediaType` function for this.

Reply:
Oh nice. Makes things easier. Added.